### PR TITLE
EKF2: Prevent jumps in vehicle yaw due to core switching

### PR DESCRIPTION
**Pull request by priseborough**
Addresses https://github.com/ArduPilot/ardupilot/issues/4679

The EKF2 publishes the amount of any yaw angle reset and the time of the reset so that the yaw controller can adjust its set-point to avoid jumps in yaw angle when the EKF performs an in-flight reset of yaw angle such as can occur when taking off out of a ground based magnetic anomaly.

The publishing of the yaw change and the time did not take into account changes in yaw angle due to a switch between EKF cores caused by IMU or other errors.

These patches ensure that a core switch also results in a the yaw step being published.

*SITL Testing*

For the first test a core switch in LOITER mode was induced by setting a very large Z gyro bias for IMU1. The yaw slews due to the fault, but when the core switch occurs, the yaw set-point jumps to match the current yaw angle and the yawing stops.

screen shot 2016-08-30 at 10 42 54 pm

For the second test the SIM_MAG_ALY_X  and SIM_MAG_ALY_Y parameters were set to -200 and 200 respectively to simulate a large ground based magnetic anomaly. This affected both cores which performed two yaw resets as the copter climbed away from the anomaly, the first being triggered by the anomaly detector, and the second the normal reset that happens at 5m when the copter switches from simple heading to 3D magnetometer fusion.

screen shot 2016-08-30 at 10 56 56 pm

### CODE REVIEW

File: libraries/AP_NavEKF2/AP_NavEKF2.h [hunk 1]
```diff
@@ -375,4 +375,14 @@ class NavEKF2
 
     // time at start of current filter update
     uint64_t imuSampleTime_us;
+
+    // used to keep track of yaw angle steps due to change of primary instance or internal ekf yaw resets
+    struct {
```
Comment (magicrub) [inline libraries/AP_NavEKF2/AP_NavEKF2.h]: Thanks for putting this into a struct :)

**Comment by OXINARF** [comment]
I had looked at this but I forgot to comment on it but I think I was seeing a potencial bug in this logic (and to honest this logic seems to complicated to me but I'll address that later).

So for anyone looking to merge this, please wait until I review it. Thanks.

**Comment by rmackay9** [comment]
I'll include this in -rc4 which I hope we can push out in about a week... assuming the review from @OXINARF and anyone else goes ok (I'll test fly it too of course).

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp [hunk 2]
```diff
@@ -1134,14 +1134,58 @@ bool NavEKF2::getHeightControlLimit(float &height) const
...
+    }
+
+    // check for a core switch and if a switch has occurred, set the yaw reset delta
+    // to the difference in yaw angle between the current and last yaw angle
+    Vector3f eulers_primary;
+    core[primary].getEulerAngles(eulers_primary);
+    if (primary != yaw_step_data.prev_instance) {
+        // the delta is the difference between the current and previous yaw
+        // This overwrites any yaw reset value recorded from an internal eff reset
+        yaw_step_data.yaw_delta += wrap_PI(eulers_primary.z - yaw_step_data.prev_yaw);
```
Comment (OXINARF) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp]: The comment says it overwrites any yaw reset value but this is doing an addition, should it just be an equal?

**Comment by OXINARF** [comment]
@priseborough Besides the comment I did inline, do we care about what the vehicles do? Right now, only Copter uses the getLastYawResetAngle and it does so in the same loop that the EKF is updated. If that was changed, for example, getLastYawResetAngle wasn't called so frequently, I can see one or two issues with this fix.

When I saw this issue my though about the way to fix it was: https://github.com/ArduPilot/ardupilot/compare/master...OXINARF:review/4757
It is just a quick draft, I haven't tested it. To me, it looks easier to follow the logic than this one, but Paul is the EKF master, so it's his decision.

**Comment by priseborough** [comment]
@OXINARF I'll have a look at your draft and then contact you to discuss the way froward on this.

**Comment by rmackay9** [comment]
@priseborough, do we think this is ready as-is for master and Copter-3.4?  I'd like to include the fix if it's safe even if there are plans to further improve the structure, etc with @OXINARF.

**Comment by priseborough** [comment]
@OXINARF  has an alternative implementation which closes a vulnerability in my implementation to not sending the correct yaw step value if there is a race condition with a yaw reset and lane switch (a low probability event). I am happy with his solution provided the error found during my review is fixed and commenting is updated. It will also need to be SITL tested.

Having said that, what is in this PR is an improvement on what we had before. 

 @OXINARF how much time do you need to fix your version and submit it for testing?

**Comment by OXINARF** [comment]
I'll do it today.

**Comment by rmackay9** [comment]
Perhaps I've jumped the gun, but I've merged this to master.  I figure it's better than what we have now and I suspect @OXINARF can build upon what Paul's done in any case.  Getting it in now takes the pressure off I think.

Thanks very much for the reviews!

**Pull request closed without merging**